src/services/createExternalRoute.ts

```diff
@@ -5,43 +5,40 @@ import { CombinePath } from '@/services/combinePath'
 import { toName, ToName } from '@/types/name'
-import { Path, toPath, ToPath } from '@/types/path'
-import { Query, toQuery, ToQuery } from '@/types/query'
 import { RouteMeta } from '@/types/register'
 import { Route } from '@/types/route'
 import { checkDuplicateParams } from '@/utilities/checkDuplicateKeys'
+import { toWithParams, ToWithParams, withParams, WithParams } from '@/services/withParams'
 
 export function createExternalRoute<
-  const THost extends string | Host,
+  const THost extends string | WithParams,
   const TName extends string | undefined = undefined,
-  const TPath extends string | Path | undefined = undefined,
-  const TQuery extends string | Query | undefined = undefined,
-  const THash extends string | Hash | undefined = undefined,
+  const TPath extends string | WithParams | undefined = undefined,
+  const TQuery extends string | WithParams | undefined = undefined,
+  const THash extends string | WithParams | undefined = undefined,
   const TMeta extends RouteMeta = RouteMeta
 >(options: CreateRouteOptions<TName, TPath, TQuery> & WithHost<THost> & WithoutParent):
-Route<ToName<TName>, ToHost<THost>, ToPath<TPath>, ToQuery<TQuery>, ToHash<THash>, TMeta>
+Route<ToName<TName>, ToWithParams<THost>, ToWithParams<TPath>, ToWithParams<TQuery>, ToWithParams<THash>, TMeta>
 
 export function createExternalRoute<
   const TParent extends Route,
   const TName extends string | undefined = undefined,
-  const TPath extends string | Path | undefined = undefined,
-  const TQuery extends string | Query | undefined = undefined,
-  const THash extends string | Hash | undefined = undefined,
+  const TPath extends string | WithParams | undefined = undefined,
+  const TQuery extends string | WithParams | undefined = undefined,
+  const THash extends string | WithParams | undefined = undefined,
   const TMeta extends RouteMeta = RouteMeta
 >(options: CreateRouteOptions<TName, TPath, TQuery> & WithoutHost & WithParent<TParent>):
-Route<ToName<TName>, Host<'', {}>, CombinePath<TParent['path'], ToPath<TPath>>, CombineQuery<TParent['query'], ToQuery<TQuery>>, CombineHash<TParent['hash'], ToHash<THash>>, CombineMeta<TMeta, TParent['meta']>>
+Route<ToName<TName>, WithParams<undefined, {}>, CombinePath<TParent['path'], ToWithParams<TPath>>, CombineQuery<TParent['query'], ToWithParams<TQuery>>, CombineHash<TParent['hash'], ToWithParams<THash>>, CombineMeta<TMeta, TParent['meta']>>
 
 export function createExternalRoute(options: CreateRouteOptions): Route {
   const id = createRouteId()
   const name = toName(options.name)
-  const path = toPath(options.path)
-  const query = toQuery(options.query)
-  const hash = toHash(options.hash)
+  const path = toWithParams(options.path)
+  const query = toWithParams(options.query)
+  const hash = toWithParams(options.hash)
   const meta = options.meta ?? {}
-  const host = isWithHost(options) ? toHost(options.host) : toHost('')
+  const host = isWithHost(options) ? toWithParams(options.host) : withParams()
```

Comment:
Why do we even need this ternary? Doesn't `toWithParams` accept undefined?

Reply:
good call, expect a couple slightly unrelated changes in this PR now